addons/survey/models/survey.py

```diff
@@ -0,0 +1,1159 @@
+           :param finished: True for completely filled survey,Falser otherwise.
+           :returns list of filtered user_input_ids.
+        '''
+        if filters:
+            domain_filter, choice = [], []
+            for filter_data in filters:
+                row_id, answer_id = filter_data['row_id'], filter_data['answer_id']
+                if row_id == 0:
+                    choice.append(answer_id)
+                else:
+                    domain_filter.extend(['|', ('value_suggested_row.id', '=', row_id), ('value_suggested.id', '=', answer_id)])
+            if choice:
+                domain_filter.insert(0, ('value_suggested.id', 'in', choice))
+            else:
+                domain_filter = domain_filter[1:]
+            user_input_lines = self.env['survey.user_input_line'].search(domain_filter)
+            filtered_input_ids = user_input_lines.mapped('user_input_ids').ids
+        else:
+            filtered_input_ids = []
+        if finished:
+            UserInput = self.env['survey.user_input']
+            if not filtered_input_ids:
+                user_inputs = UserInput.search([('survey_id', '=', survey.id)])
+            else:
+                user_inputs = UserInput.browse(filtered_input_ids)
+            return user_inputs.filtered(lambda user_input: user_input.state == 'done').ids
+        return filtered_input_ids
+
+    @api.model
+    def get_filter_display_data(self, filters):
+        '''Returns data to display current filters
+        :param filters: list of dictionary (having: row_id, answer_id)
+        :param finished: True for completely filled survey, False otherwise.
+        :returns list of dict having data to display filters.
+        '''
+        filter_display_data = []
+        if filters:
+            Surveylabel = self.env['survey.label']
+            for filter_data in filters:
+                row_id, answer_id = filter_data['row_id'], filter_data['answer_id']
+                question = Surveylabel.browse(answer_id).question_id
+                if row_id == 0:
+                    labels = Surveylabel.browse([answer_id])
+                else:
+                    labels = Surveylabel.browse([row_id, answer_id])
+                filter_display_data.append({'question_text': question.question, 'labels': [label.value for label in labels]})
+        return filter_display_data
+
+    @api.model
+    def prepare_result(self, question, current_filters=None):
+        ''' Compute statistical data for questions by counting number of vote per choice on basis of filter '''
+        current_filters = current_filters if current_filters else []
+        result_summary = {}
+
+        #Calculate and return statistics for choice
+        if question.question_type in ['simple_choice', 'multiple_choice']:
+            answers = {}
+            comments = []
+            [answers.update({label.id: {'text': label.value, 'count': 0, 'answer_id': label.id}}) for label in question.labels_ids]
```

Comment:
https://docs.python.org/2/library/stdtypes.html?highlight=dict.update#dict.update
(same idea for append)